citation_parser.py

```diff
@@ -0,0 +1,130 @@
+import re
+from collections import namedtuple
+from types import SimpleNamespace
+from nltk.tokenize import sent_tokenize
+
+Citation = namedtuple('Citation', ['text', 'prev', 'start', 'end'])
+
+class CitationParser(object):
+
+    def __init__(self):
+        self._ref_regex = '<ref.+?(/>|</ref>)'
+        self._cn_regex = '{{(C|c)itation needed\|.+?}}'
+        self._citation_regex = self._ref_regex + '|' + self._cn_regex
+        self._citations = []
+        self._sentences = []
+
+    def _find_citations(self, wikitext):
+        for ref in re.finditer(self._citation_regex, wikitext):
+            self._citations.append(Citation(text=ref.group(0), prev=wikitext[ref.start(0)-1], 
+                start=ref.start(0), end=ref.start(0)+len(ref.group(0))))
+        self._citations.reverse()
+
+    def _remove_citation_tags(self, wikitext):
+        return re.sub(self._citation_regex, '', wikitext)
+
+    def _add_successive_citation(self):
+        while self._citations:
+            if self._citations[-1].prev == '>':
+                self._sentences[-1].cite.append(self._citations.pop())
+            else:
+                break
+
+    def parse(self, wikitext):
+        """
+        This is a method for parsing wikitext to get individual sentences
+        and the citations they have.
+        
+        The algorithm goes like:
+        1) Find reference tags <ref> and citation needed templates {cn}
+           within the wikitext and store them in a queue in context order.
```

Comment:
What is "context order"?

Reply:
Oh sorry my unclear words. It means I stored them according to the sequence appears in the context.

Reply:
How about "In the order they appear in the text" then?